diff -urN a/unix/Makefile b/unix/Makefile
--- a/unix/Makefile	2020-06-23 16:19:36.588766964 +0800
+++ b/unix/Makefile	2020-06-27 17:56:29.665411455 +0800
@@ -5,7 +5,7 @@
 
 INTER=.unix
 
-HOST =
+HOST = $(CROSS_COMPILE)
 TEST_CCFLAGS =
 
 GENERATED = -DPAL_HAS_GIT_REVISION $(shell ../scripts/gengitrev)
@@ -60,7 +60,7 @@
 	LDFLAGS += -framework OpenGL
 	OBJFILES = $(filter-out ../mini_glloader.unix.o, $(TOBJFILES))
 else
-	LDFLAGS += -lGL -pthread
+	LDFLAGS += -pthread
 	OBJFILES = $(TOBJFILES)
 endif
 
diff -urN a/unix/pal_config.h b/unix/pal_config.h
--- a/unix/pal_config.h	2020-06-23 16:19:36.588766964 +0800
+++ b/unix/pal_config.h	2020-06-27 15:52:20.010582489 +0800
@@ -65,7 +65,7 @@
 #include <sys/time.h>
 
 #if SDL_VERSION_ATLEAST(2,0,0)
-# define PAL_HAS_GLSL 1
+# define PAL_HAS_GLSL 0
 #endif
 
 #endif
