diff -Nur a/Makefile b/Makefile
--- a/Makefile	2020-06-02 19:24:11.090386000 +0800
+++ b/Makefile	2020-07-07 20:57:31.814673445 +0800
@@ -256,7 +256,7 @@
 
 HOSTCC       = cc
 HOSTCXX      = c++
-HOSTCFLAGS   = -Wall -Wstrict-prototypes -O2 -fomit-frame-pointer \
+HOSTCFLAGS   = -Wall -Wstrict-prototypes -O2 -fomit-frame-pointer -Wno-format-overflow -Wno-address-of-packed-member -Wno-array-bounds -Wno-format-truncation -Wno-format-overflow \
 		$(if $(CONFIG_TOOLS_DEBUG),-g)
 HOSTCXXFLAGS = -O2
 
@@ -360,7 +360,7 @@
 KBUILD_CFLAGS   := -Wall -Wstrict-prototypes \
 		   -Wno-format-security \
 		   -fno-builtin -ffreestanding
-KBUILD_CFLAGS	+= -fshort-wchar -Werror
+KBUILD_CFLAGS	+= -fshort-wchar -Wno-address-of-packed-member
 KBUILD_AFLAGS   := -D__ASSEMBLY__
 
 # Read UBOOTRELEASE from include/config/uboot.release (if it exists)
