diff -uNr a/arch/arm64/boot/dts/rockchip/rk3399-nanopi4-common.dtsi b/arch/arm64/boot/dts/rockchip/rk3399-nanopi4-common.dtsi
--- a/arch/arm64/boot/dts/rockchip/rk3399-nanopi4-common.dtsi	2020-06-02 19:25:30.847112000 +0800
+++ b/arch/arm64/boot/dts/rockchip/rk3399-nanopi4-common.dtsi	2020-06-19 19:34:30.501776718 +0800
@@ -540,7 +540,7 @@
 };
 
 &rga {
-	status = "disabled";
+	status = "okay";
 };
 
 &display_subsystem {
